### PR TITLE
Add test plan for sub group class

### PR DESCRIPTION
**Pull request by nolmoonen**
Adds a test plan for the stream class as described in Section 4.9.1.8. of the SYCL 2020 specification.

### CODE REVIEW

**Review by bader — APPROVED** [review]
LGTM.

File: test_plans/sub_group.asciidoc [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+- `sub_group.get_max_local_range()`
++
+Returns the maximum number of items in the sub-group. Should be `sub_group_size` (see below).
+- `sub_group.leader()`
++
+Returns `true` only for the work-item within each sub-group with a local id of zero.
+
+The kernel is compiled multiple times, with the attribute `reqd_sub_group_size(dim)` for `dim` in range `1, ..., 64`. At runtime, `info::device::sub_group_sizes` is queried for the device, to obtain the supported sub-group sizes. The largest sub-group size that is smaller or equal to 64 is chosen to be `sub_group_size`. The kernel is executed with a `work_group_size` equal to `3 * sub_group_size - 1`.
+
+If no such sub-group size exists, the kernel is run without the attribute and a warning is issued. In verification, the value returned by `get_local_range` is used to calculate the expected values.
```
Comment (gmlueck) [inline test_plans/sub_group.asciidoc:44]: I think this is a question for @Pennycook.

While reviewing this test plan I was wondering about the specification for the following member functions of `sub_group`:

* `id<1> get_group_id()`
* `range<1> get_group_range()`

Why do these always return a 1-dimensional object?  Each of them returns a position of the sub-group within the containing work-group.  What if the work-group is multi-dimensional?  Wouldn't we want to return a multi-dimensional `id` and `range` in that case?

File: test_plans/sub_group.asciidoc [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+== Tests
+A kernel is defined that calls all functions below, checks their return type and records the returned value in a list of a size equal to the number of work-items times the number of functions. The kernel is called with `parallel_for(nd_range<1>(range<1>(work_group_size), range<1>(work_group_size)), [](nd_item<1> item){})`, let `sub_group` be `item.get_sub_group()`:
+
+- `sub_group.get_group_id()` and `sub_group.get_group_linear_id()`
++
+Returns the id of the sub-group within the work-group.
+- `sub_group.get_local_id()` and `sub_group.get_local_linear_id()`
++
+Returns the id work-item's position in the sub-group.
+- `sub_group.get_local_range()` and `sub_group.get_group_linear_range()`
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Fixed.

File: test_plans/sub_group.asciidoc [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
++
+Returns the size of the subgroup. Should be `sub_group_size` (see below).
+- `sub_group.get_group_range()` and `sub_group.get_local_linear_range()`
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Fixed.

File: test_plans/sub_group.asciidoc [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
++
+Returns the number of sub-groups within the work-group. Should be `3` (see below).
+- `sub_group.get_max_local_range()`
++
+Returns the maximum number of items in the sub-group. Should be `sub_group_size` (see below).
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc:37]: Well, the `get_max_local_range` should still be `sub_group_size`, right? But the `get_local_range` and `get_local_linear_range` may be smaller than `sub_group_size`, so I have modified that.

File: test_plans/sub_group.asciidoc [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+=== Backend coverage
+
+All the tests described below are not backend-specific and are performed for any SYCL backend.
+
+=== Device coverage
+
+All tests construct a test device for which conformance is assessed. All the tests described below are performed once for that test device.
+
+== Tests
+A kernel is defined that calls all functions below, checks their return type and records the returned value in a list of a size equal to the number of work-items times the number of functions. The kernel is called with `parallel_for(nd_range<1>(range<1>(work_group_size), range<1>(work_group_size)), [](nd_item<1> item){})`, let `sub_group` be `item.get_sub_group()`:
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc:21]: I suppose that would be possible but, reading the specification again, I am confused about how partitioning into sub-groups is performed. The spec states for `get_local_range()` (emphasis mine):
> Return a range representing the size of the sub-group. This size may be less than the value returned by `get_max_local_range()`, depending on the position of the sub-group within its parent work-group **and the manner in which sub-groups are constructed by the implementation**.

In the test plan, as it stands, I assume that the first $n$ work-items of a one-dimensional `nd_range` get placed in the first sub-group, the next $n$ in the second sub-group, etc. I doubt that there is a justification for an implementation to do it differently. But it is not clear to me how that would work for higher-dimensional `nd_ranges`, and if that is specified by the specification at all. This also relates to your question below https://github.com/KhronosGroup/SYCL-CTS/pull/436#discussion_r1054630603.

File: test_plans/sub_group.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc:44]: If an implementation passes these checks you suggest, this is a pretty reliable indicator that the implementation is correct, unless some serious edge-cases occur. Some of these edge-cases can be covered by, for instance, checking that all global indices appear in-order.

However, the test that forces the sub-group sizes does not have the same risk that some edge-behavior is not covered, because it is much easier to verify that a passed test means that the implementation does what it should. Furthermore, if the test fails, it is easier to tell exactly which function returned the wrong value.

I do not think that it is unreasonable to assume that the device has a sub-group size in range $[1, \ldots ,64]$. This can be increased to 128 or 256, with little cost, the program will be small and can compile in parallel.

So, I'd like to keep the current setup.

File: test_plans/sub_group.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply test_plans/sub_group.asciidoc:21]: > In the test plan, as it stands, I assume that the first $n$ work-items of a one-dimensional `nd_range` get placed in the first sub-group, the next $n$ in the second sub-group, etc.

I think this is not a valid assumption.  The spec specifically says that the number of work-items in each sub-group is implementation defined.  See this in [section 3.9.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels):

> The size and number of sub-groups is implementation-defined and may differ for each kernel, and different devices may make different guarantees with respect to how sub-groups within a work-group are scheduled. The maximum number of work-items in any sub-group in a kernel is based on a combination of the kernel and its dispatch dimensions. The size of any sub-group in the dispatch is between 1 and this maximum sub-group size, and the size of an individual sub-group is invariant for the duration of a kernel’s execution.

The more I think about it, the more I think that my proposed checks in [this comment](https://github.com/KhronosGroup/SYCL-CTS/pull/436#discussion_r1054621832) are the proper way to test the `sub_group` class.  We need to be careful that the CTS follows the letter of the spec and does not make assumptions about what a "reasonable implementation" should do.

This discussion has made me wonder whether the `[sycl::reqd_sub_group_size]` attribute should provide some stronger guarantee.  For example, maybe that attribute should require the work-group size to be evenly divisible by the required sub-group size (in each dimension), and then the spec could guarantee that every sub-group has exactly the requested size.  Tagging @Pennycook for comment on this.  Note, though, that the spec does *not* provide this guarantee now, so we should not test for this in the CTS.

File: test_plans/sub_group.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply test_plans/sub_group.asciidoc:44]: Adding some more thoughts here about why it is not a good idea to rely on the `[[sycl::reqd_sub_group_size]]` attribute when testing the `sub_group` class:

* The implementation may behave quite differently when this attribute is present.  The CTS should cover both cases: both when this attribute is present and when it is not present.  Therefore, I think the CTS should include a test of *just* the `sub_group` class (without the attribute).  If we want to test additional semantics that are specific to the attribute, we should write a separate test for the `[[sycl::reqd_sub_group_size]]` attribute.

* Some implementations (e.g. hipSYCL) may not implement the C++ attributes.  It would be good if the CTS could tell whether such an implementation is conformant to the `sub_group` semantics without giving a failure due to missing attribute support.

File: test_plans/sub_group.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: You raise valid points, in the other comment thread as well. I agree that fully relying on `sycl::reqd_sub_group_size` for the sub-group tests is flawed. As such, I have updated the test plan to include this test when the sub-group size is not known beforehand.

Do you think it is best to also drop the part of the test with `sycl::reqd_sub_group_size`? If I understand it correctly, the expected values for `subgroup.get_group_id`, `subgroup.get_local_id`, etc. cannot be calculated, even if the sub-group size is known beforehand due to the fact that the implementation may choose how sub-groups are constructed.

File: test_plans/sub_group.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Please see [this conversation](https://github.com/KhronosGroup/SYCL-CTS/pull/436#discussion_r1059971667), I have added these checks you suggest.

File: test_plans/sub_group.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: To illustrate the latter point:
```
work-group range: (2,3)
sub-group size:   2

division might be:
[(0,0),(0,1)],[(0,2),(1,0)],[(1,1),(1,2)]
with local_ids: 0, 1, 0, 1, 0, 1

or, it might be:
[(0,0),(0,1)],[(0,2)],[(1,0),(1,1)],[(1,2)]
with local_ids: 0, 1, 0, 0, 1, 0
```
The purpose of the test that enforced the sub-group size, is that the expected values of the member functions of the sub-group class can be calculated. Since this seems not to be the case, in my opinion it is best to leave this test out of the plan.

File: test_plans/sub_group.asciidoc [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+== Tests
+Two tests are defined, depending on whether the test has knowledge of the actual sub-group size. In some cases, this may not be enforceable and be decided by the compiler.
+
+=== With unknown sub-group size
+This test is always executed. A kernel is defined that calls all functions below, checks their return type and records the returned value in a list of a size equal to the number of work-items times the number of functions. This list is initialized such that it can be asserted that each element is written to.
+
+The kernel is called with `parallel_for(nd_range<3>(global_range, local_range), [](nd_item<1> item){})`, where `local_range` is `1023` in every dimension to likely contain multiple sub-groups and likely to contain an incomplete sub-group. `local_range` is decreased depending on `info::device::max_work_group_size` and `info::device::max_work_item_sizes`. `global_range` is `range<3>(2, 2, 2) * local_range`. The work-items index the list of results with `item.get_global_linear_id()`. Let `sub_group` be `item.get_sub_group()`:
```
Comment (gmlueck) [inline test_plans/sub_group.asciidoc:26]:
```suggestion
The kernel is called with `parallel_for(nd_range<3>(global_range, local_range), [](nd_item<3> item){})`, where `local_range` is `1023` in every dimension to likely contain multiple sub-groups and likely to contain an incomplete sub-group. `local_range` is decreased depending on `info::device::max_work_group_size` and `info::device::max_work_item_sizes`. `global_range` is `range<3>(2, 2, 2) * local_range`. The work-items index the list of results with `item.get_global_linear_id()`. Let `sub_group` be `item.get_sub_group()`:
```

I assume this was just a typo.

File: test_plans/sub_group.asciidoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply test_plans/sub_group.asciidoc]: BTW, I captured some of the open questions that have arisen from this test plan here: https://gitlab.khronos.org/sycl/Specification/-/issues/638

[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Okay, I have removed this part from the test plan.

File: test_plans/sub_group.asciidoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: This part is now removed from the test plan.

File: test_plans/sub_group.asciidoc (hunk 6 above)
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Yes, I have fixed it.

File: test_plans/sub_group.asciidoc [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+- `sub_group.leader()`
+- `item.get_group_linear_id()`
+
+The following tests are performed for each work-item:
+
+- `sub_group.get_group_id()[0]` is equal to `sub_group.get_group_linear_id()` and within `sub_group.get_group_range()[0]`.
+- `sub_group.get_local_id()[0]` is equal to `sub_group.get_local_linear_id()` and within `sub_group.get_local_range()[0]`.
+- `sub_group.get_local_range()[0]` is equal to `sub_group.get_local_linear_range()` and is smaller or equal to `sub_group.get_max_local_range()[0]`.
+- `sub_group.get_group_range()[0]` is equal to `sub_group.get_group_linear_range()`.
+- Calculate a global index for each work-item using `sub_group.get_group_id()`, `sub_group.get_local_id()`, and `item.get_group_linear_id()`. Each global index should be unique.
```
[the remark this answers is not in the document]
Comment (nolmoonen) [reply test_plans/sub_group.asciidoc]: Okay, I see the problem. Even if the global index calculated by the test does not need to be equal to the actual global work-item index (just some abstract unique value), the calculation can still be problematic as there may be multiple incomplete sub-groups within a work-group.

I have dropped the global index test and replaced it with a (more specific) version of the test you suggest.

**Review by gmlueck — APPROVED** [review]

File: test_plans/sub_group.asciidoc (hunk 1 above)
Comment (keryell) [reply test_plans/sub_group.asciidoc]: This comes from OpenCL https://github.com/KhronosGroup/OpenCL-Docs/blob/main/ext/cl_khr_subgroups.asciidoc

**Comment by fraggamuffin** [comment]
LGTM
